examples/training/classification/imagenet/basic_training.py

```diff
@@ -197,49 +217,115 @@ def augment(img, label):
         **eval(FLAGS.model_kwargs),
     )
 
-
 """
-Next, we pick an optimizer. Here we use Adam with a constant learning rate.
-Note that learning rate will decrease over time due to the ReduceLROnPlateau callback.
+Optional LR schedule with cosine decay instead of ReduceLROnPlateau
+TODO: Replace with Core Keras LRWarmup when it's released. This is a temporary solution.
 """
 
 
-optimizer = optimizers.SGD(
-    learning_rate=INITIAL_LEARNING_RATE, momentum=0.9, global_clipnorm=10
+def lr_warmup_cosine_decay(
+    global_step, warmup_steps, hold=0, total_steps=0, start_lr=0.0, target_lr=1e-3
+):
+    # Cosine decay
+    # There is no tf.pi :(
+    learning_rate = (
+        0.5
+        * target_lr
+        * (
+            1
+            + tf.cos(
+                tf.constant(np.pi)
+                * (global_step - warmup_steps - hold)
+                / float(total_steps - warmup_steps - hold)
+            )
+        )
+    )
+    warmup_lr = target_lr * (global_step / warmup_steps)
+
+    if hold > 0:
+        learning_rate = tf.where(
+            global_step > warmup_steps + hold, learning_rate, target_lr
+        )
+
+    learning_rate = tf.where(global_step < warmup_steps, warmup_lr, learning_rate)
+    return learning_rate
+
+
+class WarmUpCosineDecay(keras.optimizers.schedules.LearningRateSchedule):
+    def __init__(self, start_lr, target_lr, warmup_steps, total_steps, hold):
+        super().__init__()
+        self.start_lr = start_lr
+        self.target_lr = target_lr
+        self.warmup_steps = warmup_steps
+        self.total_steps = total_steps
+        self.hold = hold
+
+    def __call__(self, step):
+        lr = lr_warmup_cosine_decay(
+            global_step=step,
+            total_steps=self.total_steps,
+            warmup_steps=self.warmup_steps,
+            start_lr=self.start_lr,
+            target_lr=self.target_lr,
+            hold=self.hold,
+        )
+
+        return tf.where(step > self.total_steps, 0.0, lr, name="learning_rate")
+
+
+# If batched
+total_steps = len(train_ds) * FLAGS.epochs
+warmup_steps = int(FLAGS.warmup_steps_percentage * total_steps)
```

Comment:
Both of these need to be multiplied by 1/100 (10% == 0.1)
Re-started the training run :-)

Reply:
🤦‍♂️ I usually write them as `0..1` floats instead of integers, and it seemed like most other do integers instead, so I used them in this proposal, totally mentally forgetting that in the call below. My bad!

Reply:
No worries at all!